### PR TITLE
DNF 2.0: Add system upgrade

### PR DESCRIPTION
**Pull request by Conan-Kudo**
This PR is identical to #66 (and indeed has the same commits), except that the commit altering the spec file has been rediffed against the current master and does not do more than add the system upgrade subpackage.

This should be acceptable for inclusion provided that everything else in #66 was fine.

### CODE REVIEW

File: dnf-plugins-extras.spec [hunk 1]
```diff
@@ -257,6 +257,47 @@ Obsoletes:      dnf-plugins-extras-snapper <= 0.0.4-2
...
+Provides:       system-upgrade = %{?epoch:%{epoch}:}%{version}-%{release}
+
+Obsoletes:	fedup < 0.9.4
+Obsoletes:	dnf-plugin-system-upgrade < 0.10
+Obsoletes:	python3-dnf-plugin-system-upgrade < 0.10
+
+BuildRequires:  pkgconfig(systemd)
+BuildRequires:  systemd
+BuildRequires:  python3-systemd
+%{?systemd_requires}
```
Comment (ignatenkobrain) [inline dnf-plugins-extras.spec]: missing Req: python3-systemd

File: dnf-plugins-extras.spec [hunk 2]
```diff
@@ -257,6 +257,47 @@ Obsoletes:      dnf-plugins-extras-snapper <= 0.0.4-2
...
+%{?python_provide:%python_provide python2-%{name}-system-upgrade}
+
+Obsoletes:	fedup < 0.9.4
+Obsoletes:	dnf-plugin-system-upgrade < 0.10
+Obsoletes:	python2-dnf-plugin-system-upgrade < 0.10
+
+BuildRequires:  pkgconfig(systemd)
+BuildRequires:  systemd
+BuildRequires:  python2-systemd
+%{?system_requires}
```
Comment (ignatenkobrain) [inline dnf-plugins-extras.spec]: missing Req: python3-systemd

Comment (Conan-Kudo) [inline dnf-plugins-extras.spec]: What are you talking about, they are there. It's above the BuildRequires. Also, I think you mean "python2-systemd" here.

File: dnf-plugins-extras.spec (hunk 1 above)
Comment (Conan-Kudo) [inline dnf-plugins-extras.spec]: Look at line 282.

File: dnf-plugins-extras.spec (hunk 2 above)
Comment (ignatenkobrain) [inline dnf-plugins-extras.spec]: oh

**Comment by mhatina** [comment]
I have tried system-upgrade on f24 (upgrading to f25).
`# dnf clean cache all`
`# dnf system-upgrade download --refresh --releasever 25 --exclude kernel\*`
(sometimes I get `Error: Failed to synchronize cache for repo 'fedora'`)
`# dnf system-upgrade reboot`

Result was:
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe37: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe39: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe3a: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe3b: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe3c: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe3d: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe3e: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe3f: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe40: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe41: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe42: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe43: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe44: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe45: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe46: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe47: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe48: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe49: 30528/140395134830336`
`warning: rpmdb: BDB2053 Freeing read locks for locker 0xe4a: 30528/140395134830336`
`Traceback (most recent call last):`
` File "/bin/dnf", line 58, in <module>`
`  main.user_main(sys.argv[1:], exit_code=True)`
` File "/usr/lib/python3.5/site-packages/dnf/cli/main.py", line 174, in user_main`
`    errcode = main(args)`
` File "/usr/lib/python3.5/site-packages/dnf/cli/main.py", line 61, in main`
`    return _main(base, args)`
`  File "/usr/lib/python3.5/site-packages/dnf/cli/main.py", line 99, in _main`
`    return cli_run(cli, base)`
`  File "/usr/lib/python3.5/site-packages/dnf/cli/main.py", line 115, in cli_run`
`    cli.run()`
`  File "/usr/lib/python3.5/site-packages/dnf/cli/cli.py", line 951, in run`
`    return self.command.run()`
`  File "/home/mhatina/dnf-plugins-extras/plugins/system_upgrade.py", line 335, in run`
`    self._call_sub("run", self.opts)`
`  File "/home/mhatina/dnf-plugins-extras/plugins/system_upgrade.py", line 343, in _call_sub`
`    subfunc(*args)`
`  File "/home/mhatina/dnf-plugins-extras/plugins/system_upgrade.py", line 433, in run_reboot`
`    REBOOT_REQUESTED_ID)`
`  File "/home/mhatina/dnf-plugins-extras/plugins/system_upgrade.py", line 329, in log_status`
`    DNF_VERSION=dnf.const.VERSION)`
`  File "/usr/lib64/python3.5/site-packages/systemd/journal.py", line 446, in send`
`    args.extend(_make_line(key, val) for key, val in kwargs.items())`
`  File "/usr/lib64/python3.5/site-packages/systemd/journal.py", line 446, in <genexpr>`
`    args.extend(_make_line(key, val) for key, val in kwargs.items())`
`  File "/usr/lib64/python3.5/site-packages/systemd/journal.py", line 400, in _make_line`
`    return field + '=' + value`
`TypeError: Can't convert 'NoneType' object to str implicitly`

Then I have tried the same with sudo:
`$ sudo dnf clean cache all`
`$ sudo dnf system-upgrade download --refresh --releasever 25 --exclude kernel\\*`
(sometimes I get `Error: Failed to synchronize cache for repo 'fedora'`)
`$ sudo dnf system-upgrade reboot`

Result:
`[Errno 17] File exists: '/var/lib/dnf/system-upgrade' -> '/system-update'`
So I have `rm -rf`ed `/var/lib/dnf/system-upgrade` and `/system-update`, and tried again with sudo.

Result:
See traceback above.


And I get this everytime I run dnf:
`Failed loading plugin: show_leaves`
`Failed loading plugin: repomanage`
`Failed loading plugin: migrate`
`Failed loading plugin: versionlock`
`Failed loading plugin: torproxy`
`Failed loading plugin: rpm_conf`
`Failed loading plugin: tracer`
`Failed loading plugin: debug`
`Failed loading plugin: repograph`
`Failed loading plugin: local`
`Failed loading plugin: kickstart`
`Failed loading plugin: snapper`
`Failed loading plugin: repoclosure`
`Failed loading plugin: leaves`
`Loaded plugins: builddep, config-manager, copr, debuginfo-install, download, generate_completion_cache, ghost, needs-restarting, noroot, playground, reposync, system-upgrade`
`Ghost plugin: loaded (with CLI)`

**Comment by stepasm** [comment]
PR that fix offline remove of non-running kernel: [libhif PR/210](https://github.com/rpm-software-management/libhif/pull/210)

IDK why it failed... can you provide more info on how did you test it?

Plugins are most probably not loaded because dnf-plugins-extras-system-upgrade requires dnf > 2.0 and you have installed plugins that depend on dnf-1.0, which should be fixed in [dnf-plugins-extras PR/70](https://github.com/rpm-software-management/dnf-plugins-extras/pull/70). So are you sure that you used new dnf with new dnf-plugins-extras-system-upgrade?

BTW Anaconda still uses dnf-1.0 so u need to remove anaconda as u are not able to provide both dnf-1.0 and dnf-2.0.

**Comment by Conan-Kudo** [comment]
@mhatina I've rebased the PR against current master.

**Comment by mhatina** [comment]
I have successfully tested system-upgrade. Problem was that I was missing some dirs/files (commands for their creation are in specfile).

**Comment by mhatina** [comment]
@dnf-bot r+

**Pull request closed without merging**